app/models/agents/ftpsite_agent.rb

```diff
@@ -32,42 +50,64 @@ class FtpsiteAgent < Agent
     MD
 
     def working?
-      event_created_within?(interpolated['expected_update_period_in_days']) && !recent_error_logs?
+      if interpolated['mode'] == 'read'
+        event_created_within?(interpolated['expected_update_period_in_days']) && !recent_error_logs?
+      else
+        received_event_without_error?
+      end
     end
 
     def default_options
       {
+        'mode' => 'read',
         'expected_update_period_in_days' => "1",
         'url' => "ftp://example.org/pub/releases/",
         'patterns' => [
           'foo-*.tar.gz',
         ],
         'after' => Time.now.iso8601,
+        'force_encoding' => '',
+        'filename' => '',
+        'data_path' => '$.data'
       }
     end
 
     def validate_options
       # Check for required fields
       begin
-        url = options['url']
+        url = interpolated['url']
         String === url or raise
         uri = URI(url)
         URI::FTP === uri or raise
-        errors.add(:base, "url must end with a slash") unless uri.path.end_with?('/')
+        errors.add(:base, "url must end with a slash") if uri.path.present? && !uri.path.end_with?('/')
       rescue
         errors.add(:base, "url must be a valid FTP URL")
       end
 
-      patterns = options['patterns']
-      case patterns
-      when Array
-        if patterns.empty?
-          errors.add(:base, "patterns must not be empty")
+      if options['mode'].blank? || !['read', 'write'].include?(options['mode'])
```

Comment:
The Agents would even need to somehow keep track of their version. In this case it would be simple: set mode to `read` when it is not present, like the migration does. But yes, it will probably get painful soon :smile: